crates/futures/src/lib.rs

```diff
@@ -289,9 +316,108 @@ fn _future_to_promise(future: Box<dyn Future<Item = JsValue, Error = JsValue>>)
         Waiting(Arc<Package>),
     }
 
+    #[cfg(target_feature = "atomics")]
+    struct Waker {
+        array: Vec<i32>,
+        notified: AtomicBool,
+    };
+
+    #[cfg(target_feature = "atomics")]
+    impl Waker {
+        fn new(array: Vec<i32>, notified: bool) -> Self {
+            Waker {
+                array,
+                notified: AtomicBool::new(notified),
+            }
+        }
+    }
+
+    #[cfg(target_feature = "atomics")]
+    impl Notify for Waker {
+        fn notify(&self, id: usize) {
+            console_log!("Waker notify");
+            if !self.notified.swap(true, Ordering::SeqCst) {
+                console_log!("Waker, inside if");
+                let memory_buffer = wasm_bindgen::memory()
+                .dyn_into::<WebAssembly::Memory>()
+                .expect("Should cast a memory to WebAssembly::Memory")
+                .buffer();
+
+                let array_location = self.array.as_ptr() as u32 / 4;
+                let array = Int32Array::new(&memory_buffer)
+                    .subarray(array_location, array_location + self.array.len() as u32);
+
+                let _ = Atomics::notify(&array, id as u32);
+            }
+        }
+    }
+
+    #[cfg(target_feature = "atomics")]
+    fn poll_again(package: Arc<Package>, id: usize) {
+        console_log!("poll_again called");
+        let me = match package.notified.replace(State::Notified) {
+            // we need to schedule polling to resume, so keep going
+            State::Waiting(me) => {
+                console_log!("poll_again Waiting");
+                me
+            },
+
+            // we were already notified, and were just notified again;
+            // having now coalesced the notifications we return as it's
+            // still someone else's job to process this
+            State::Notified => {
+                console_log!("poll_again Notified");
+                return;
+            },
+
+            // the future was previously being polled, and we've just
+            // switched it to the "you're notified" state. We don't have
+            // access to the future as it's being polled, so the future
+            // polling process later sees this notification and will
+            // continue polling. For us, though, there's nothing else to do,
+            // so we bail out.
+            // later see
+            State::Polling => {
+                console_log!("poll_again Polling");
+                return;
+            },
+        };
+
+        let memory_buffer = wasm_bindgen::memory()
+            .dyn_into::<WebAssembly::Memory>()
+            .expect("Should cast a memory to WebAssembly::Memory")
+            .buffer();
+
+        let array_location = package.waker.array.as_ptr() as u32 / 4;
+        let array = Int32Array::new(&memory_buffer)
+            .subarray(array_location, array_location + package.waker.array.len() as u32);
```

Comment:
I think it's safe to avoid the `subarray` here and just use the `Int32Array` with a larger offset that is the pointer that we're waiting on.